src/trace/log.zig

```diff
@@ -1,366 +1,464 @@
-    }
-
-    pub fn spawn(self: Self) void {
-        switch (self) {
-            .standard => |logger| {
-                logger.spawn();
-            },
-            .noop, .test_logger => {},
-        }
-    }
-
-    pub fn deinit(self: Self) void {
-        switch (self) {
-            .standard => |logger| {
-                logger.deinit();
-            },
-            .noop, .test_logger => {},
-        }
-    }
-
-    pub fn field(self: Self, name: []const u8, value: anytype) Entry {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                return logger.field(name, value);
-            },
-            .noop => {
-                return .noop;
-            },
-        }
-    }
-
-    pub fn infof(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.infof(fmt, args);
-            },
-            .noop => {},
-        }
-    }
-
-    pub fn debugf(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.debugf(fmt, args);
-            },
-            .noop => {},
+/// A ScopedLogger could either be:
+/// - A StandardErrLogger
+/// - A TestingLogger
+pub fn ScopedLogger(comptime scope: ?[]const u8) type {
+    return union(LogKind) {
+        const Self = @This();
+        standard: *ChannelPrintLogger,
+        test_logger: *DirectPrintLogger,
+        noop: void,
+
+        /// Can be used in tests to minimize the amount of logging during tests.
+        pub const TEST_DEFAULT_LEVEL: Level = .warn;
+
+        pub fn unscoped(self: Self) Logger {
+            return switch (self) {
+                .standard => |logger| .{ .standard = logger },
+                .test_logger => |logger| .{ .test_logger = logger },
+                .noop => .noop,
+            };
         }
-    }
-
-    pub fn warnf(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.warnf(fmt, args);
-            },
-            .noop => {},
+        pub fn withScope(self: Self, comptime new_scope: []const u8) ScopedLogger(new_scope) {
+            return switch (self) {
+                .standard => |logger| .{ .standard = logger },
+                .test_logger => |logger| .{ .test_logger = logger },
+                .noop => .noop,
+            };
         }
-    }
 
-    pub fn errf(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.errf(fmt, args);
-            },
-            .noop => {},
+        pub fn deinit(self: *const Self) void {
+            switch (self.*) {
+                .standard => |*logger| {
+                    logger.*.deinit();
```

Comment:
Is this idiomatic? meaning `var.*.func()`

Reply:
Cc: @0xNineteen

Reply:
yeah good catch - we shouldnt need that - https://github.com/Syndica/sig/pull/277#discussion_r1790775768